mantidproject/mantid
# Reflectometry GUI restructuring 7: processing

**Description of work.**

Ongoing work to integrate the `JobTreeView` into the `ISIS Reflectometry` GUI. This PR is being merged to a feature branch because not all functionality is implemented yet.

This PR implements processing of runs in the `ISIS Reflectometry` GUI. It includes the following main changes:
 - A new `Batch` model in the `Reduction` subfolder brings together the entire reduction configuration.
- Each `Row`/`Group` now stores its state (complete, error etc.) via a new base class, `Item`
- A new `BatchJobRunner`, which is called by the `BatchPresenter`, sets up the reduction algorithms for each row using properties from the reduction configuration. It sets up the queue in a `BatchAlgorithmRunner`, which is owned by the `BatchView`.
- The `BatchPresenter` executes the `BatchAlgorithmRunner` via the `BatchView`.
- The `BatchAlgorithmRunner` has been modified to emit signals to the view when the individual algorithms complete or fail.
- A cancel method has been added to the `BatchAlgorithmRunner`. The cancel flag will be called from the main thread while the batch is running on the child thread, so the flag is protected by a mutex. Processing will be stopped once the current algorithm has finished executing.
- A new `ConfiguredAlgorithm` class has been added to replace the original `std::pair` used to link the algorithm to its properties. An override of this is used by `BatchJobRunner` to pass data about the row/group to update when an algorithm completes.
- A new `TransmissionRunPair` class has been added to replace the previous pair of strings. This supports multiple input run numbers for each transmission input, which are summed prior to reduction.

**Report to:** nobody

**To test:**
- Open the `ISIS Reflectometry` interface.
- Set the instrument to `INTER`
- Add a group to the table, and in the `Run(s)` cell of the row enter `13460` and for `Angle` enter `0.5`.
- Click `Process`. The row should be highlighted to indicate success.
- Check that row state is reset when output workspaces are deleted from the ADS (but not when renamed)
- Re-process and check that row state is reset when you edit cell values in the table.
- Re-process and check that row state is reset when you change settings on the Event, Experiment or Instrument tabs.
- Check that any settings you change on the table and settings tabs get reflected in the input properties to the `ReflectometryLoadAndProcess` algorithm history on the `IvsQ_binned_` named workspace (this is the final output).
- Some code has been removed which worked out output workspace names for the reduction. These are now taken from the reduction algorithm.
- Note that there are some TODOs deliberately left in regarding injecting a model into the presenters to specify defaults. This will be done in a future PR.



*There is no associated issue.*

*This does not require release notes* because it only concerns under-the-hood changes

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
